### PR TITLE
fix: correct PaddleOCR-VL and PP-DocLayout on Metal

### LINKED ISSUE #177: PaddleOCR-VL on Metal: non-deterministic results, most regions decode to empty/garbage (CPU is correct)
**Body**:

## Summary

Running PaddleOCR-VL-1.5 through `oar-ocr-vl` 0.9.1 on Apple Silicon with the `metal` feature produces unstable recognition: the same input image yields different outputs across runs, most regions decode to empty strings or runaway garbage (e.g. `\(^{2000000000000000000.0000…`), and overall output shrinks to a fraction of the expected content. The identical pipeline on CPU produces complete, correct results.

## Environment

- oar-ocr-vl 0.9.1 (crates.io), features = ["metal"]
- Model: PaddlePaddle/PaddleOCR-VL-1.5 (HF safetensors, BF16), PP-DocLayoutV3_safetensors
- macOS (Darwin 25.5), Apple Silicon, 8 GB RAM
- rustc 1.97.1, release build

## Reproduction

```rust
let layout = PpDocLayout::from_dir(layout_dir, Device::Cpu)?; // see note 1
let vl = PaddleOcrVl::from_dir(model_dir, metal_device)?;
let parser = DocParser::new(&vl);
let md = parser.parse_to_markdown(&layout, img)?; // img: 3306x4164 bilevel book scan
```

Run the same image several times:

- Run 1: partial output (top half of page), ends with runaway `\(^{20000000…` in a superscript/formula region (~1.2 KB total)
- Run 2 (same binary, same input): different, smaller output (~0.5 KB), most text regions empty
- With `OAR_VL_DTYPE=f32` and/or `OAR_VL_DISABLE_METAL_SDPA=1`: all regions decode empty (0 usable output)
- **CPU device: complete and correct output every run** (~2.1 KB, matches the page)

Input characteristics that may matter: large (3306×4164) bilevel (pure black/white) scanned page.

## Note 1: PP-DocLayout on Metal fails outright

`PpDocLayout` cannot run on Metal at all — the sinusoidal position embedding is built as an f64 tensor (`pp_doclayout/encoder.rs`, `Tensor::from_vec(vec<f64>, …)`) and Candle's Metal backend does not support F64, so inference fails with `inference failed in model 'PP-DocLayout': build position embedding`. Building the embedding in f64 on CPU (or in f32) before casting would fix Metal for the layout stage. This is why the repro above pins layout to CPU.

## Expected

Metal results should match CPU (within fp tolerance), and repeated runs on the same input should be deterministic.

## Workaround we ship

We currently pin the whole VL pipeline to CPU on macOS (correct but ~3-4× slower). Happy to test patches on this hardware.

---

### PR DESCRIPTION
**Pull request by GreatV**
Fixes #177.

PP-DocLayout could not run on Metal at all, and PaddleOCR-VL produced unstable, mostly-empty output there while CPU was correct. Investigating with the real checkpoints on an M1 Pro turned up four distinct bugs.

## 1. PP-DocLayout fails outright on Metal

`sine_position_embedding` built its f64 table with `Tensor::from_vec(vec![0f64; …], …, device)`, materializing an F64 tensor directly on the target device — Metal has no F64 support. This is exactly the reporter's diagnosis, and reverting the fix reproduces their error string verbatim (`inference failed in model 'PP-DocLayout': build position embedding`).

Now built and downcast on CPU, then uploaded, so CPU results stay bit-identical. This was the only F64 tensor construction in `pp_doclayout`.

## 2. candle's Metal reduce kernel makes every detection garbage

With #1 fixed the detector *ran* but returned nonsense — 10 detections vs CPU's 41, low scores, boxes outside the image. Bisecting stage by stage: backbone ✅, encoder ✅, encoder head ✅, topk indices bit-identical ✅, V3 mask branch ✅, decoder layer 1 self-attention ✅ — then `cross` diverged hard (CPU −444.32 vs Metal +1370.70) and the final layer emitted 7500/7500 non-finite logits.

Root cause is upstream: in `candle-metal-kernels-0.11.0/src/metal_src/reduce.metal`, `get_strided_index_t` only specializes ranks 1–4 (`case 5:`/`case 6:` are commented out) and falls back to `get_strided_idx_fallback<4>`, which peels the outermost dimension as if it were fastest-varying and then re-peels dims 3..0, never touching dim 4. Measured boundary:

| rank | non-last axis | last axis |
|---|---|---|
| 2–4 | correct | correct |
| 5–6 | **wrong** | correct |

The deformable blend hits this with `sum(4).sum(3)` over a rank-6 tensor. It now folds `levels` and `points` into one axis and reduces a rank-3 view — same maths, on paths every backend gets right. I swept the crate for other rank ≥ 5 reductions; this was the only one.

After: **41 detections vs 41, worst |Δbox| 0.0004 px, worst |Δscore| 1e-6.**

Worth reporting upstream to candle.

## 3. Left-padded rows decode to NaN and never terminate

`create_left_padding_mask` filled padded positions with `-inf`. Combined with the causal mask, a query *inside* the padding region can only reach padding keys, so that row is entirely `-inf`, its softmax is NaN, and `0 * NaN` spreads the NaN across the sequence. Measured: exactly 50% of the prefill hidden state non-finite at batch=2 — one whole row.

The result decoded to garbage and never emitted EOS, burning the full `max_new_tokens` budget (4096 steps for ~30 bytes of output). This is the "runaway garbage" in the issue report.

Now uses a large finite negative, matching what `create_generation_mask` already did. **225s → 3.96s** on the reproducer.

## 4. Vision encoder blends images together

`VisionModel::forward` packed all images into one sequence (`cat` then `unsqueeze(0)`) and ran the encoder with **no attention mask**, so every patch of one image attended to the others. Two regions came back as a blend of each other's text (`绿洲仕格维` + `打浦路252935` → `绿浦路2栋维`).

Now encodes one image at a time — what upstream's block-diagonal `cu_seqlens` mask computes — leaving the single-image path byte-identical.

Also fixed alongside: the decode-step MRoPE position tensor is `(axis, batch, 1)` but was filled batch-major, transposing it and giving each row another row's positions. The two layouts coincide only at `batch_size == 1`.

## Secondary

- `free_device_memory` now reports Metal headroom via `recommendedMaxWorkingSetSize - currentAllocatedSize`, so the memory-aware chunked vision attention from #147 can actually engage there (it previously saw `None` and always took the full-matrix path). Verified against hardware: 12123 MiB → 11099 MiB after a 1 GiB allocation.
- Metal vision attention routes through the fused SDPA kernel (`head_dim` 72 is supported), avoiding the materialized `[b, heads, seq, seq]` scores and their F32 softmax round-trip. Modest speed win (~3%) but a real cut in peak memory, which matters for the 8 GB machine in the report. Metal-only; the eager path stays byte-stable for CUDA.
- Fixed a pre-existing failing test, `test_metal_causal_query_longer_than_kv_matches_eager`, which asserted behaviour `create_causal_mask` no longer has.

## Tests

New regression coverage, all inline next to the code: the folded rank-3 reduction vs the rank-6 form; a Metal-only canary that fails once candle's reduce is fixed, so the workaround gets removed; the padding fill staying finite in F32/BF16/F16; and the axis-major MRoPE buffer across its 3- and 4-axis users.

`pp_doclayout::model::tests::metal_matches_cpu` is an opt-in CPU/Metal parity check, `#[ignore]`d like the PyTorch reference test beside it:

```sh
OAR_PP_DOCLAYOUT_DIR=$PWD/models/PP-DocLayoutV3_safetensors \
OAR_PP_DOCLAYOUT_IMAGE=$PWD/page.png \
    cargo test --release -p oar-ocr-vl --features metal -- --ignored metal_matches_cpu
```

`cargo test -p oar-ocr-vl --features metal` (189) and the full workspace suite pass; clippy clean on both feature sets.

## Adjacent fixes found in review

The batch-major decode position buffer (bug 4's companion) was present verbatim in two more backends and is fixed here too: MinerU2.5's 3-axis version, and HunyuanOCR's 4-axis one, which sits in the `else` of `batch_size == 1` and so is wrong every time it executes. Both are reachable from their public batched `generate`. The three now share `decode_position_buffer` and one test.

Also: the fused Metal SDPA short-circuit initially sat ahead of the sequence threshold, which silently made `OAR_VL_ATTN_FULL_SEQ_THRESHOLD=0` a no-op on Metal despite being documented to force chunked attention. It is now gated on the threshold.

## Note for reviewers

Bugs 3 and 4 only trigger when several regions are recognized in one batch, which `DocParser` does not currently do. They are real defects in the public `PaddleOcrVl::generate` API, which documents multi-image batching — and they are a prerequisite for any future work that enables recognition batching, since it would otherwise be wrong from the start.

Separately: on this hardware decode already runs at ~8.2 ms/token against a ~9 ms/token memory-bandwidth floor, so the remaining time is the vision encoder (~550 ms/region). Batching *decode* across regions measured slower, not faster, because the batch runs until the longest sequence finishes. Batching only the vision encoder looks like the real lever, but that is a design change and is not in this PR.

### CODE REVIEW

**Comment by GreatV** [comment]
Confirmed and fixed in cf7bbc7 — good catch.

`-1e9` casts to `-inf` in F16 (BF16/F32 are fine), so on F16 the finite fill had no effect and the all-masked-row NaN came straight back. F16 is reachable both ways noted: `OAR_VL_DTYPE=f16` and `select_dtype`'s fallback on CUDA devices without BF16 kernels; all three of PaddleOCR-VL, MinerU and HunyuanOCR call `create_left_padding_mask`.

The fill is now per-dtype (`-1e4` for F16), kept well inside F16's 65504 range so `score + fill` stays finite too — F16's finite minimum would have overflowed again once real logits were added. `create_generation_mask` had the same latent issue and got the same fix.

Added `padding_mask_is_finite_in_every_dtype` over F32/BF16/F16; verified it fails on the old fill with `F16: padding-region row is entirely non-finite: [-inf, ...]`. End-to-end: BF16 output byte-identical (4612 bytes), `OAR_VL_DTYPE=f16` now decodes correctly.
